File: ios/Airadar/Models/MilestoneTier.swift
import SwiftUI

/// The nine visible tiers, plus one hidden one above them — run the way a
/// real airline's elite tiers are: promotion happens on legs flown OR
/// distance flown SINCE THE LAST RANK-UP, whichever gets there first, and
/// both counters reset to zero the moment that happens — a long-haul
/// traveller and a short-hop commuter both climb, just by different
/// arithmetic, and neither a single mega-flight nor a burst of short hops
/// vaults more than one tier at a time. (The lifetime flight total shown
/// elsewhere — My's own stats panel — is a separate, never-reset count;
/// it just isn't what decides a rank-up.)
///
/// Porcelain (751+ legs) is a secret: nothing before it hints it exists, and
/// reaching it is rare enough that it also carries a sequence number — the
/// Nth traveller ever to get there — awarded by the backend, never the
/// client (`BackendClient.claimPorcelain`). It's the true ceiling this
/// ladder used to call Rainbow — renamed, and split off from what used to
/// be one enormous top band (Obsidian, then Rainbow) into three narrower
/// ones (Amber, Silk, Porcelain) so the climb from Ruby onward still feels
/// like a series of rank-ups, not one long plateau.
///
/// 首飛 (First Flight) has its own artwork but isn't a rung on this ladder —
/// it's a one-off "flew at all" stamp, unlocked at leg 1 regardless of tier,
/// parked for now (its own place in the UI is a later pass).
enum MilestoneTier: Int, CaseIterable, Comparable {
    case blackIron, bronze, silver, gold, platinum, diamond, ruby, amber, silk, porcelain

    static func < (a: MilestoneTier, b: MilestoneTier) -> Bool { a.rawValue < b.rawValue }

    /// The average distance of a leg, used to convert a tier's leg-width
    /// into an equivalent km budget — the traveller's own figure.
    static let averageKmPerLeg = 1157

    /// The lifetime leg-count band this tier owns — legs 1-10 are Black
    /// Iron, 11-20 Bronze, and so on, however a traveller actually gets
    /// there (some ranks up early on distance, see `TierStanding.compute`).
    var legBand: ClosedRange<Int> {
        switch self {
        case .blackIron: 1...10
        case .bronze: 11...20
        case .silver: 21...30
        case .gold: 31...50
        case .platinum: 51...75
        case .diamond: 76...100
        case .ruby: 101...250
        case .amber: 251...450
        case .silk: 451...750
        case .porcelain: 751...Int.max
        }
    }

    /// The band's width in legs — Porcelain, the true ceiling, has none.
    var legsInBand: Int { self == .porcelain ? Int.max : legBand.upperBound - legBand.lowerBound + 1 }

    /// The band's width converted to km at the average leg length — the
    /// budget that resets to zero every time a rank-up happens.
    var kmBudget: Int { legsInBand == Int.max ? Int.max : legsInBand * Self.averageKmPerLeg }

    var next: MilestoneTier? {
        let ordered = Self.allCases.sorted()
        guard let i = ordered.firstIndex(of: self), i + 1 < ordered.count else { return nil }
        return ordered[i + 1]
    }

    var name: String {
        switch self {
        case .blackIron: String(localized: "Black Iron")
        case .bronze: String(localized: "Bronze")
        case .silver: String(localized: "Silver")
        case .gold: String(localized: "Gold")
        case .platinum: String(localized: "Platinum")
        case .diamond: String(localized: "Diamond")
        case .ruby: String(localized: "Ruby")
        case .amber: String(localized: "Amber")
        case .silk: String(localized: "Silk")
        case .porcelain: String(localized: "Porcelain")
        }
    }

    /// Only ever shown for the tier with no next — today just Porcelain.
    var tagline: String {
        switch self {
        case .blackIron: String(localized: "Every journey starts here.")
        case .bronze: String(localized: "Ten flights and counting.")
        case .silver: String(localized: "A quarter-century in the air.")
        case .gold: String(localized: "Fifty flights strong.")
        case .platinum: String(localized: "Triple digits — a real habit now.")
        case .diamond: String(localized: "Diamond-clear dedication.")
        case .ruby: String(localized: "Two-fifty, and still climbing.")
        case .amber: String(localized: "Preserved in flight, one leg at a time.")
        case .silk: String(localized: "Smooth as the old trade routes.")
        case .porcelain: String(localized: "A secret worth finding.")
        }
    }

    /// The real artwork's file key — `tier-<key>-logo.png` under
    /// `Resources/Tiers`. Placeholder gradients only kick in if a file is
    /// ever missing (a tier added here before its art arrives, say).
    var artKey: String {
        switch self {
        case .blackIron: "blackiron"
        case .bronze: "bronze"
        case .silver: "silver"
        case .gold: "gold"
        case .platinum: "platinum"
        case .diamond: "diamond"
        case .ruby: "ruby"
        case .amber: "amber"
        case .silk: "silk"
        case .porcelain: "porcelain"
        }
    }

    /// Placeholder centre → edge palette — only ever seen if `artKey`'s
    /// file is missing from the bundle.
    var gradient: [Color] {
        switch self {
        case .blackIron: [Color(red: 0.24, green: 0.25, blue: 0.27), Color(red: 0.09, green: 0.09, blue: 0.11)]
        case .bronze: [Color(red: 0.88, green: 0.66, blue: 0.47), Color(red: 0.54, green: 0.29, blue: 0.17)]
        case .silver: [Color(red: 0.96, green: 0.96, blue: 0.97), Color(red: 0.62, green: 0.64, blue: 0.67)]
        case .gold: [Color(red: 1.00, green: 0.95, blue: 0.69), Color(red: 0.79, green: 0.58, blue: 0.10)]
        case .platinum: [Color(red: 0.93, green: 0.95, blue: 0.97), Color(red: 0.66, green: 0.72, blue: 0.77)]
        case .diamond: [Color(red: 0.92, green: 0.99, blue: 1.00), Color(red: 0.37, green: 0.75, blue: 0.88)]
        case .ruby: [Color(red: 1.00, green: 0.54, blue: 0.54), Color(red: 0.48, green: 0.05, blue: 0.12)]
        case .amber: [Color(red: 1.00, green: 0.82, blue: 0.40), Color(red: 0.72, green: 0.42, blue: 0.05)]
        case .silk: [Color(red: 0.98, green: 0.90, blue: 0.86), Color(red: 0.80, green: 0.63, blue: 0.58)]
        case .porcelain: [Color(red: 0.98, green: 0.99, blue: 1.00), Color(red: 0.55, green: 0.65, blue: 0.82)]
        }
    }

    var rimColor: Color {
        switch self {
        case .blackIron: Color(red: 0.36, green: 0.38, blue: 0.41)
        case .bronze: Color(red: 0.79, green: 0.48, blue: 0.24)
        case .silver: Color(red: 0.79, green: 0.80, blue: 0.82)
        case .gold: Color(red: 0.95, green: 0.79, blue: 0.30)
        case .platinum: Color(red: 0.81, green: 0.88, blue: 0.93)
        case .diamond: Color(red: 0.74, green: 0.94, blue: 1.00)
        case .ruby: Color(red: 0.85, green: 0.12, blue: 0.25)
        case .amber: Color(red: 0.88, green: 0.56, blue: 0.10)
        case .silk: Color(red: 0.90, green: 0.75, blue: 0.70)
        case .porcelain: Color(red: 0.30, green: 0.42, blue: 0.68)
        }
    }
}

/// Where a traveller actually stands: the tier reached after walking every
/// completed flight in order, and how many legs and how many km have
/// counted toward the CURRENT tier so far — both reset to zero at every
/// rank-up, which is what makes "legs or km, whichever first" a real race
/// tier by tier instead of a one-time fuse.
struct TierStanding {
    let tier: MilestoneTier
    let legsIntoTier: Int
    let kmIntoTier: Int

    var legsToNext: Int? { tier.next.map { _ in max(0, tier.legsInBand - legsIntoTier) } }
    var kmToNext: Int? { tier.next.map { _ in max(0, tier.kmBudget - kmIntoTier) } }

    /// Every completed flight, oldest first, feeds both counters; whichever
    /// hits the current tier's budget first promotes — same as an airline
    /// combining segments and miles toward the same rank-up.
    static func compute(for flights: [Flight]) -> TierStanding {
        let ordered = flights.sorted { ($0.departureInstant ?? .distantPast) < ($1.departureInstant ?? .distantPast) }
        var tier = MilestoneTier.blackIron
        var legs = 0, km = 0
        for f in ordered {
            legs += 1
            km += f.distanceKm
            while tier.next != nil, legs >= tier.legsInBand || km >= tier.kmBudget {
                tier = tier.next!
                legs = 0; km = 0
            }
        }
        return TierStanding(tier: tier, legsIntoTier: legs, kmIntoTier: km)
    }
}

/// The medallion — real per-tier artwork when it's in the bundle
/// (`Resources/Tiers/tier-<key>-logo.png`), a placeholder gradient if not.
struct TierBadgeView: View {
    let tier: MilestoneTier
    var size: CGFloat = 28

    private static var cache: [String: UIImage] = [:]
    private var art: UIImage? {
        if let cached = Self.cache[tier.artKey] { return cached }
        guard let url = Bundle.main.url(forResource: "tier-\(tier.artKey)-logo", withExtension: "png"),
              let data = try? Data(contentsOf: url), let image = UIImage(data: data) else { return nil }
        Self.cache[tier.artKey] = image
        return image
    }

    var body: some View {
        Group {
            if let art {
                Image(uiImage: art).resizable().scaledToFit()
            } else {
                ZStack {
                    Circle().fill(RadialGradient(colors: tier.gradient, center: UnitPoint(x: 0.35, y: 0.32), startRadius: 0, endRadius: size * 0.75))
                    Circle().strokeBorder(tier.rimColor, lineWidth: max(1.5, size * 0.07))
                    Image(systemName: "airplane").font(.system(size: size * 0.42, weight: .semibold))
                        .foregroundStyle(.white).rotationEffect(.degrees(45))
                }
            }
        }
        .frame(width: size, height: size)
    }
}

extension Color {
    /// A plain linear blend toward `other` — used to fade the route-progress
    /// plane and its flown line from one tier's colour into the next's as
    /// it moves along, rather than snapping from one to the other.
    func mixed(with other: Color, fraction: Double) -> Color {
        let f = min(1, max(0, fraction))
        var r1: CGFloat = 0, g1: CGFloat = 0, b1: CGFloat = 0, a1: CGFloat = 0
        var r2: CGFloat = 0, g2: CGFloat = 0, b2: CGFloat = 0, a2: CGFloat = 0
        UIColor(self).getRed(&r1, green: &g1, blue: &b1, alpha: &a1)
        UIColor(other).getRed(&r2, green: &g2, blue: &b2, alpha: &a2)
        return Color(red: r1 + (r2 - r1) * f, green: g1 + (g2 - g1) * f, blue: b1 + (b2 - b1) * f, opacity: a1 + (a2 - a1) * f)
    }
}

File: app/src/main/java/com/airadar/app/data/MilestoneTier.kt
package com.airadar.app.data

import androidx.compose.ui.graphics.Color
import java.time.Instant

/**
 * The nine visible tiers, plus one hidden one above them — run the way a
 * real airline's elite tiers are: promotion happens on legs flown OR
 * distance flown SINCE THE LAST RANK-UP, whichever gets there first, and
 * both counters reset to zero the moment that happens — a long-haul
 * traveller and a short-hop commuter both climb, just by different
 * arithmetic, and neither a single mega-flight nor a burst of short hops
 * vaults more than one tier at a time. (The lifetime flight total shown
 * elsewhere — My's own stats panel — is a separate, never-reset count; it
 * just isn't what decides a rank-up.)
 *
 * Porcelain (751+ legs) is a secret: nothing before it hints it exists, and
 * reaching it is rare enough that it also carries a sequence number — the
 * Nth traveller ever to get there — awarded by the backend, never the
 * client ([BackendClient.claimPorcelain]). It's the true ceiling this ladder
 * used to call Rainbow -- renamed, and split off from what used to be one
 * enormous top band (Obsidian, then Rainbow) into three narrower ones
 * (Amber, Silk, Porcelain).
 *
 * Declared in ladder order -- Kotlin enums compare by declaration (ordinal)
 * order, matching iOS's own `rawValue` ordering, so `<`/`compareTo` need no
 * extra code here.
 */
enum class MilestoneTier {
    BLACK_IRON, BRONZE, SILVER, GOLD, PLATINUM, DIAMOND, RUBY, AMBER, SILK, PORCELAIN;

    /** The lifetime leg-count band this tier owns -- legs 1-10 are Black
     * Iron, 11-20 Bronze, and so on, however a traveller actually gets
     * there (some ranks up early on distance, see [TierStanding.compute]). */
    val legBand: IntRange
        get() = when (this) {
            BLACK_IRON -> 1..10
            BRONZE -> 11..20
            SILVER -> 21..30
            GOLD -> 31..50
            PLATINUM -> 51..75
            DIAMOND -> 76..100
            RUBY -> 101..250
            AMBER -> 251..450
            SILK -> 451..750
            PORCELAIN -> 751..Int.MAX_VALUE
        }

    /** The band's width in legs -- Porcelain, the true ceiling, has none. */
    val legsInBand: Int get() = if (this == PORCELAIN) Int.MAX_VALUE else legBand.last - legBand.first + 1

    /** The band's width converted to km at the average leg length -- the
     * budget that resets to zero every time a rank-up happens. */
    val kmBudget: Int get() = if (legsInBand == Int.MAX_VALUE) Int.MAX_VALUE else legsInBand * AVERAGE_KM_PER_LEG

    val next: MilestoneTier?
        get() {
            val ordered = entries
            val i = ordered.indexOf(this)
            return ordered.getOrNull(i + 1)
        }

    /** English for now, matching the rest of the app -- the Chinese names
     * (黑鐵/青銅/白銀/黃金/白金/鑽石/紅寶石/黑曜石/彩虹) wait for the real
     * localization pass. */
    val displayName: String
        get() = when (this) {
            BLACK_IRON -> tr("Black Iron")
            BRONZE -> tr("Bronze")
            SILVER -> tr("Silver")
            GOLD -> tr("Gold")
            PLATINUM -> tr("Platinum")
            DIAMOND -> tr("Diamond")
            RUBY -> tr("Ruby")
            AMBER -> tr("Amber")
            SILK -> tr("Silk")
            PORCELAIN -> tr("Porcelain")
        }

    /** Only ever shown for the tier with no next -- today just Porcelain. */
    val tagline: String
        get() = when (this) {
            BLACK_IRON -> tr("Every journey starts here.")
            BRONZE -> tr("Ten flights and counting.")
            SILVER -> tr("A quarter-century in the air.")
            GOLD -> tr("Fifty flights strong.")
            PLATINUM -> tr("Triple digits — a real habit now.")
            DIAMOND -> tr("Diamond-clear dedication.")
            RUBY -> tr("Two-fifty, and still climbing.")
            AMBER -> tr("Preserved in flight, one leg at a time.")
            SILK -> tr("Smooth as the old trade routes.")
            PORCELAIN -> tr("A secret worth finding.")
        }

    /** The real artwork's file key -- `tier_<key>_logo` under
     * `res/drawable-nodpi`. Placeholder gradients only kick in if a
     * drawable is ever missing (a tier added here before its art arrives, say). */
    val artKey: String
        get() = when (this) {
            BLACK_IRON -> "blackiron"
            BRONZE -> "bronze"
            SILVER -> "silver"
            GOLD -> "gold"
            PLATINUM -> "platinum"
            DIAMOND -> "diamond"
            RUBY -> "ruby"
            AMBER -> "amber"
            SILK -> "silk"
            PORCELAIN -> "porcelain"
        }

    /** Placeholder centre -> edge palette -- only ever seen if [artKey]'s
     * drawable is missing from resources. */
    val gradient: List<Color>
        get() = when (this) {
            BLACK_IRON -> listOf(Color(0xFF3E4045), Color(0xFF17181C))
            BRONZE -> listOf(Color(0xFFE1A978), Color(0xFF8A4A2B))
            SILVER -> listOf(Color(0xFFF5F5F7), Color(0xFF9EA3AB))
            GOLD -> listOf(Color(0xFFFFF2B0), Color(0xFFCA941A))
            PLATINUM -> listOf(Color(0xFFEDF2F7), Color(0xFFA8B8C4))
            DIAMOND -> listOf(Color(0xFFEBFDFF), Color(0xFF5EBFE0))
            RUBY -> listOf(Color(0xFFFF8A8A), Color(0xFF7A0D1F))
            AMBER -> listOf(Color(0xFFFFD166), Color(0xFFB86B0D))
            SILK -> listOf(Color(0xFFFAE6DB), Color(0xFFCCA093))
            PORCELAIN -> listOf(Color(0xFFFAFCFF), Color(0xFF8CA6D1))
        }

    val rimColor: Color
        get() = when (this) {
            BLACK_IRON -> Color(0xFF5C6169)
            BRONZE -> Color(0xFFC97A3D)
            SILVER -> Color(0xFFC9CBD0)
            GOLD -> Color(0xFFF2C94D)
            PLATINUM -> Color(0xFFCFE1EC)
            DIAMOND -> Color(0xFFBCF0FF)
            RUBY -> Color(0xFFD91F40)
            AMBER -> Color(0xFFE08F1A)
            SILK -> Color(0xFFE6BFB3)
            PORCELAIN -> Color(0xFF4D6BAD)
        }

    companion object {
        /** The average distance of a leg, used to convert a tier's leg-width
         * into an equivalent km budget -- the traveller's own figure. */
        const val AVERAGE_KM_PER_LEG = 1157
    }
}

/**
 * Where a traveller actually stands: the tier reached after walking every
 * completed flight in order, and how many legs and how many km have
 * counted toward the CURRENT tier so far -- both reset to zero at every
 * rank-up, which is what makes "legs or km, whichever first" a real race
 * tier by tier instead of a one-time fuse.
 */
data class TierStanding(val tier: MilestoneTier, val legsIntoTier: Int, val kmIntoTier: Int) {

    val legsToNext: Int? get() = tier.next?.let { (tier.legsInBand - legsIntoTier).coerceAtLeast(0) }
    val kmToNext: Int? get() = tier.next?.let { (tier.kmBudget - kmIntoTier).coerceAtLeast(0) }

    companion object {
        /** Every completed flight, oldest first, feeds both counters;
         * whichever hits the current tier's budget first promotes -- same
         * as an airline combining segments and miles toward the same rank-up. */
        fun compute(flights: List<Flight>): TierStanding {
            val ordered = flights.sortedBy { it.departureInstant ?: Instant.MIN }
            var tier = MilestoneTier.BLACK_IRON
            var legs = 0
            var km = 0
            for (f in ordered) {
                legs += 1
                km += f.distanceKm
                while (true) {
                    val next = tier.next ?: break
                    if (legs >= tier.legsInBand || km >= tier.kmBudget) {
                        tier = next
                        legs = 0; km = 0
                    } else break
                }
            }
            return TierStanding(tier, legs, km)
        }
    }
}

/** A plain linear blend toward [other] -- used to fade the route-progress
 * plane and its flown line from one tier's colour into the next's as it
 * moves along, rather than snapping from one to the other. */
fun Color.mixed(with: Color, fraction: Float): Color {
    val f = fraction.coerceIn(0f, 1f)
    return Color(
        red = red + (with.red - red) * f,
        green = green + (with.green - green) * f,
        blue = blue + (with.blue - blue) * f,
        alpha = alpha + (with.alpha - alpha) * f
    )
}
